src/main/java/frc/robot/subsystems/drivetrain/BaseSwerveSubsystem.java

```diff
@@ -235,6 +249,23 @@ public void setDrivePowers(double xPower) {
         setDrivePowers(xPower, 0.0, 0.0, true);
     }
 
+    /**
+     * Sets the swerve module states of this subsystem from provided field-centric
+     * swerve drive powers, locking the swerve to a given heading.
+     * 
+     * @param xPower The power [-1.0, 1.0] in the x (forward) direction.
+     * @param yPower The power [-1.0, 1.0] in the y (left) direction.
+     * @param heading The `Rotation2d` angle to lock the swerve to.
+     * @param relative Whether to use relative powers instead of field-oriented control.
+     */
+    public void setDrivePowersWithHeadingLock(double xPower, double yPower, Rotation2d heading, boolean relative) {
+        Rotation2d currentRotation = getRobotPosition().getRotation();
+        double error = MathUtil.angleModulus(heading.minus(currentRotation).getRadians());
+        double turnSpeed = thetaController.calculate(error);
```

Comment:
Oh another thing, we seem to be using a `ProfiledPIDController` which is profiled in units of omega (radians / second) and feeding its output directly to turnpower (which should be percent output between `[-1.0, 1.0]`)? Should we scale the calculated omega by dividing it by `MAX_OMEGA` first?

Reply:
this explains a lot .

yheah lol

Reply:
sorry i've been stuck in the "swerve uses velocities!!" mindset